internal/api/controllers/private/highlevelConnectionStatus.go

```diff
@@ -0,0 +1,247 @@
+			connectionStatus = "connected"
+		} else {
+			connectionStatus = "disconnected"
+		}
+
+		responses = append(responses, formatConnectionResponse(nil, nil, host.RHCClientID, orgId, []string{host.ID}, string(RecipientType_directConnect), connectionStatus))
+	}
+
+	return responses, nil
+}
+
+func getSatelliteStatus(ctx echo.Context, client connectors.CloudConnectorClient, sourceClient sources.SourcesConnector, orgId OrgId, hostDetails []inventory.HostDetails) ([]RecipientWithConnectionInfo, error) {
+	hostsGroupedBySatellite := groupHostsBySatellite(hostDetails)
+
+	hostsGroupedBySatellite = getSourceInfo(ctx, hostsGroupedBySatellite, sourceClient)
+
+	responses, err := createSatelliteConnectionResponses(ctx, hostsGroupedBySatellite, client, orgId)
+	if err != nil {
+		utils.GetLogFromEcho(ctx).Error("error occured creating satellite connection response")
+		return nil, ctx.NoContent(http.StatusInternalServerError)
+	}
+
+	return responses, nil
+}
+
+func groupHostsBySatellite(hostDetails []inventory.HostDetails) map[string]*rhcSatellite {
+	hostsGroupedBySatellite := make(map[string]*rhcSatellite)
+
+	for _, host := range hostDetails {
+		satInstanceAndOrg := *host.SatelliteInstanceID + *host.SatelliteOrgID
+		_, exists := hostsGroupedBySatellite[satInstanceAndOrg]
+
+		if exists {
+			hostsGroupedBySatellite[satInstanceAndOrg].Hosts = append(hostsGroupedBySatellite[satInstanceAndOrg].Hosts, host.ID)
+		} else {
+			hostsGroupedBySatellite[satInstanceAndOrg] = &rhcSatellite{
+				SatelliteInstanceID: *host.SatelliteInstanceID,
+				SatelliteOrgID:      *host.SatelliteOrgID,
+				SatelliteVersion:    *host.SatelliteVersion,
+				Hosts:               []string{host.ID},
+			}
+		}
+	}
+
+	return hostsGroupedBySatellite
+}
+
+func getSourceInfo(ctx echo.Context, hostsGroupedBySatellite map[string]*rhcSatellite, sourceClient sources.SourcesConnector) map[string]*rhcSatellite {
+	for i, satellite := range hostsGroupedBySatellite {
+		result, err := sourceClient.GetSourceConnectionDetails(ctx.Request().Context(), satellite.SatelliteInstanceID)
+
+		if err != nil {
+			utils.GetLogFromEcho(ctx).Errorf("Sources data could not be found for SatelliteID %s Error: %s", satellite.SatelliteInstanceID, err)
+		} else {
+			hostsGroupedBySatellite[i].SourceID = result.ID
+			hostsGroupedBySatellite[i].RhcClientID = result.RhcID
+			hostsGroupedBySatellite[i].SourceAvailabilityStatus = result.AvailabilityStatus
+		}
+	}
+
+	return hostsGroupedBySatellite
+}
+
+func createSatelliteConnectionResponses(ctx echo.Context, hostsGroupedBySatellite map[string]*rhcSatellite, cloudConnector connectors.CloudConnectorClient, orgId OrgId) ([]RecipientWithConnectionInfo, error) {
+	responses := []RecipientWithConnectionInfo{}
+
+	for _, satellite := range hostsGroupedBySatellite {
+		if satellite.RhcClientID != nil {
+			status, err := cloudConnector.GetConnectionStatus(ctx.Request().Context(), satellite.SatelliteOrgID, satellite.SatelliteInstanceID)
```

Comment:
Cloud-Connector will need the `RhcClientID` in order to check the connection status.  In this case (hosts connected through a satellite), is the `satellite.SatelliteInstanceID` really the rhc-client-id for the connection?  Or do we need to call sources to map the SatelliteInstanceID to and rhc-client-id?

Oh, it looks like `getSourceInfo()` sets the RhcClientID on the host entry.  `getSourceInfo()` is getting called before this method.  So I think `RhcClientID` is what needs to be passed to `cloudConnector.GetConnectionStatus()` here.

There are so many ids to keep track of around here.  :)

Reply:
Ah!  Good catch!  Yes that's right, thank you! :)